### ISSUE TITLE
4.x: Base CakePHP Cache classes on PSR-6

### ISSUE THREAD

**Issue by inoas** [cakephp/cakephp#9507]
This is a (multiple allowed):
* [x] enhancement

* CakePHP Version: 4.x

Ref: http://www.php-fig.org/psr/psr-6/
Ref: https://packagist.org/packages/symfony/cache

Just in case the core team considers this worth time and efforts.

**Comment by ravage84** [comment]
Just want to mention PSR-16 ([draft](http://www.php-fig.org/psr/#index-by-status)):

> PSR-6 solves this problem already, but in a rather formal and verbose way for what the most simple use cases need. This simpler approach aims to build a standardized layer of simplicity on top of the existing PSR-6 interfaces.

https://github.com/php-fig/fig-standards/blob/master/proposed/simplecache.md

**Comment by dakota** [comment]
I'd much rather see a PSR-16 implementation, but I'm not too much of a fan of either.

**Comment by josegonzalez** [comment]
PSR-16 was accepted, so if someone wants to spend some time working on the PR, that'd be *awesome*.

**Comment by ADmad** [comment]
> Cache Misses - A cache miss will return null and therefore detecting if one stored null is not possible.

This is one quirk I noticed. PSR-16 dictates returning `null` on cache miss while our cache engines return `false`.

**Comment by josegonzalez** [comment]
Feck, can we do that for 4.0? Not sure if that would break our promise on not breaking BC /cc @markstory .

**Comment by ADmad** [comment]
I wonder how that's actually handled in implementations since for e.g `Memcached::get()` return `false` on failure not `null`.

**Comment by josegonzalez** [comment]
I'm guessing internally they implement PSR-6 so they cache CacheItemInterface objects.

**Comment by markstory** [comment]
Switching false to null is within the scope of breaking changes I think are ok for 4.0. Improving overall community compatibility is a good reason for a small BC break, and it would be in a major release too.

**Comment by prisis** [comment]
How about if you use https://github.com/php-cache/cache?

It has support for psr-6 and psr-16

**Comment by markstory** [comment]
@prisis That is something we could explore. It could be a good place to source PSR6 adapters, and the PSR16 bridge adapter.

**Comment by prisis** [comment]
If you need some help or have questions, you can ask me

**Comment by markstory** [comment]
@prisis Will do. One gap I'm seeing is the lack of a wincache adapter. Would you be open to adding support for this adapter? You'd be able to do CI with Appveyor.

**Comment by prisis** [comment]
@markstory yes, I can add a wincache adapter, just need so free time :D, the appveyor support is in the backlog.
